File: task/agents.py
# task/agents.py
"""
AI agent strategy implementations for the RPS Social Cognition Task.

All four social agents play at the same strategic level (k ≈ 1–2 in CHASE).
The RNG control plays uniformly at random. Agents are based on a simplified
CHASE model (Buergi et al., 2026) with fixed sophistication — no adaptive
belief updating on the agent side, but calibrated noise to ensure human-like
behavior that passes a Turing test.
"""

import numpy as np
from typing import List, Optional
from task.config import AGENT_STRATEGY_LEVELS, AGENT_NOISE, ACTIONS


class BaseAgent:
    """Abstract base for all agents."""

    def __init__(self, agent_id: str, seed: Optional[int] = None):
        self.agent_id    = agent_id
        self.rng         = np.random.default_rng(seed)
        self.action_history: List[int]  = []
        self.opp_history: List[int]     = []

    def choose(self, opponent_history: List[int]) -> int:
        raise NotImplementedError

    def update(self, own_action: int, opp_action: int) -> None:
        self.action_history.append(own_action)
        self.opp_history.append(opp_action)

    def reset(self) -> None:
        self.action_history = []
        self.opp_history    = []


class RandomAgent(BaseAgent):
    """
    Level-0 / RNG control.
    Chooses uniformly at random — no strategy, no mind to model.
    Participants are told there is no opponent; a generic icon is shown.
    """

    def choose(self, opponent_history: List[int]) -> int:
        return int(self.rng.choice(ACTIONS))


class CHASEAgent(BaseAgent):
    """
    Fixed-level CHASE agent (Buergi et al., 2026).

    Implements A1 (action-frequency tracking) + A2 (recursive reasoning)
    at a fixed sophistication level k, with calibrated noise gamma so the
    agent is indistinguishable from human opponents in a Turing test.

    Parameters
    ----------
    level : int
        Fixed strategic level k (1 or 2).
    alpha : float
        Learning rate for action-frequency tracker (attractions).
    beta : float
        Softmax noise for action selection.
    noise : float
        Additional epsilon-greedy noise to humanise behavior.
    """

    def __init__(
        self,
        agent_id:  str,
        level:     int   = 1,
        alpha:     float = 0.3,
        beta:      float = 3.0,
        noise:     float = 0.15,
        seed:      Optional[int] = None,
    ):
        super().__init__(agent_id, seed)
        self.level  = level
        self.alpha  = alpha
        self.beta   = beta
        self.noise  = noise  # Epsilon-greedy noise probability

        # Action-frequency attractions (A1), initialised uniformly
        self.attractions = np.ones(len(ACTIONS)) / len(ACTIONS)

        # Payoff matrix: payoff[i][j] = 1 if action i beats action j, -1 if loses, 0 if draw
        # Actions indexed 0,1,2 = Rock, Paper, Scissors
        self._payoff = np.array([
            [ 0, -1,  1],   # Rock:     draws Rock, loses to Paper, beats Scissors
            [ 1,  0, -1],   # Paper:    beats Rock, draws Paper, loses to Scissors
            [-1,  1,  0],   # Scissors: loses to Rock, beats Paper, draws Scissors
        ], dtype=float)

    def _softmax(self, values: np.ndarray) -> np.ndarray:
        v = self.beta * values
        v -= v.max()
        e = np.exp(v)
        return e / e.sum()

    def _level0_probs(self) -> np.ndarray:
        """P(a | k=0): softmax over current attractions."""
        return self._softmax(self.attractions)

    def _best_respond(self, opp_probs: np.ndarray) -> np.ndarray:
        """
        Given a distribution over opponent actions, compute the
        best-response distribution (one step of recursive reasoning).
        Returns softmax over expected payoffs.
        """
        ev = self._payoff @ opp_probs
        return self._softmax(ev)

    def _compute_action_probs(self) -> np.ndarray:
        """Compute P(a | k=self.level) by applying k steps of recursion."""
        probs = self._level0_probs()
        for _ in range(self.level):
            probs = self._best_respond(probs)
        return probs

    def choose(self, opponent_history: List[int]) -> int:
        """
        Select an action given the opponent's history.
        Uses the agent's own history (stored internally) to update attractions.
        """
        probs = self._compute_action_probs()

        # Epsilon-greedy noise: with probability `noise`, choose uniformly
        if self.rng.random() < self.noise:
            action_idx = int(self.rng.choice(len(ACTIONS)))
        else:
            action_idx = int(self.rng.choice(len(ACTIONS), p=probs))

        return ACTIONS[action_idx]

    def update(self, own_action: int, opp_action: int) -> None:
        """Update action-frequency attractions after each trial."""
        super().update(own_action, opp_action)

        # Delta-rule update on own past actions (A1)
        action_idx = ACTIONS.index(own_action)
        indicator  = np.zeros(len(ACTIONS))
        indicator[action_idx] = 1.0
        self.attractions += self.alpha * (indicator - self.attractions)

    def reset(self) -> None:
        super().reset()
        self.attractions = np.ones(len(ACTIONS)) / len(ACTIONS)


def build_agent(agent_id: str, seed: Optional[int] = None) -> BaseAgent:
    """
    Factory: return the appropriate agent object for a given agent_id.
    All social agents (A–D) use CHASE at the same strategic level so that
    mentalizing differences are driven by prior connection, not difficulty.
    """
    if agent_id == "rng":
        return RandomAgent(agent_id, seed=seed)

    level = AGENT_STRATEGY_LEVELS[agent_id]
    noise = AGENT_NOISE[agent_id]
    return CHASEAgent(agent_id, level=level, noise=noise, seed=seed)

File: rps-social-cognition-2/frontend/src/agents.ts
/**
 * Agent strategy logic for the RPS task.
 * Ported from rps_task.html — mulberry32 seeded RNG, RandomAgent, CHASEAgent.
 */

// ── Seeded RNG (mulberry32) ──────────────────────────────────────────────────

function mkRng(seed: number | null): () => number {
  if (seed === null) return () => Math.random();
  let s = seed;
  return () => {
    s |= 0;
    s = (s + 0x6d2b79f5) | 0;
    let t = Math.imul(s ^ (s >>> 15), 1 | s);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function weightedChoice(rng: () => number, probs: number[]): number {
  const r = rng();
  let c = 0;
  for (let i = 0; i < probs.length; i++) {
    c += probs[i]!;
    if (r < c) return i;
  }
  return probs.length - 1;
}

// ── Constants ────────────────────────────────────────────────────────────────

const ACTIONS: readonly [1, 2, 3] = [1, 2, 3];

// ── Agent interface ──────────────────────────────────────────────────────────

export interface Agent {
  readonly id: string;
  choose(): number;
  update(own: number, opp: number): void;
  reset(): void;
}

// ── RandomAgent ──────────────────────────────────────────────────────────────

class RandomAgent implements Agent {
  readonly id: string;
  private rng: () => number;

  constructor(id: string, seed: number | null) {
    this.id = id;
    this.rng = mkRng(seed);
  }

  choose(): number {
    return ACTIONS[Math.floor(this.rng() * 3)]!;
  }

  update(_own: number, _opp: number): void {}
  reset(): void {}
}

// ── CHASEAgent ───────────────────────────────────────────────────────────────
// Level-k recursive best-response with softmax + delta-rule attraction updates.

interface CHASEConfig {
  level?: number;
  alpha?: number;
  beta?: number;
  noise?: number;
  seed?: number | null;
}

class CHASEAgent implements Agent {
  readonly id: string;
  private level: number;
  private alpha: number;
  private beta: number;
  private noise: number;
  private rng: () => number;
  private attr: number[];

  // Row-player payoff matrix: payoff[myAction][theirAction]
  // 0=Rock, 1=Paper, 2=Scissors (0-indexed internally)
  private readonly payoff = [
    [0, -1, 1],
    [1, 0, -1],
    [-1, 1, 0],
  ];

  constructor(id: string, config: CHASEConfig = {}) {
    this.id = id;
    this.level = config.level ?? 1;
    this.alpha = config.alpha ?? 0.3;
    this.beta = config.beta ?? 3.0;
    this.noise = config.noise ?? 0.15;
    this.rng = mkRng(config.seed ?? null);
    this.attr = [1 / 3, 1 / 3, 1 / 3];
  }

  private softmax(v: number[]): number[] {
    const s = v.map((x) => this.beta * x);
    const m = Math.max(...s);
    const e = s.map((x) => Math.exp(x - m));
    const sum = e.reduce((a, b) => a + b, 0);
    return e.map((x) => x / sum);
  }

  /** Level-0 belief: softmax over own action attractions. */
  private l0(): number[] {
    return this.softmax(this.attr);
  }

  /** Best-response distribution against an opponent mixed strategy. */
  private br(op: number[]): number[] {
    const ev = this.payoff.map((row) =>
      row.reduce((s, v, j) => s + v * op[j]!, 0),
    );
    return this.softmax(ev);
  }

  /** Compute action probabilities at configured level. */
  private probs(): number[] {
    let p = this.l0();
    for (let k = 0; k < this.level; k++) p = this.br(p);
    return p;
  }

  choose(): number {
    const p = this.probs();
    const idx =
      this.rng() < this.noise
        ? Math.floor(this.rng() * 3)
        : weightedChoice(this.rng, p);
    return ACTIONS[idx]!;
  }

  update(own: number, _opp: number): void {
    const idx = ACTIONS.indexOf(own as 1 | 2 | 3);
    this.attr = this.attr.map(
      (a, i) => a + this.alpha * ((i === idx ? 1 : 0) - a),
    );
  }

  reset(): void {
    this.attr = [1 / 3, 1 / 3, 1 / 3];
  }
}

// ── Agent config (all social agents at same CHASE level) ─────────────────────

export const AGENT_CONFIG: Record<
  string,
  { level: number; noise: number | null }
> = {
  agent_a: { level: 1, noise: 0.15 },
  agent_b: { level: 1, noise: 0.15 },
  agent_c: { level: 1, noise: 0.15 },
  agent_d: { level: 1, noise: 0.15 },
  rng: { level: 0, noise: null },
};

// ── Factory ──────────────────────────────────────────────────────────────────

export function buildAgent(agentId: string, seed: number | null): Agent {
  if (agentId === "rng") return new RandomAgent(agentId, seed);
  const cfg = AGENT_CONFIG[agentId];
  if (!cfg) return new RandomAgent(agentId, seed);
  return new CHASEAgent(agentId, {
    level: cfg.level,
    noise: cfg.noise ?? 0.15,
    seed,
  });
}
